### PR TITLE
Testing set intersect and union

### PR DESCRIPTION
**Pull request by Divesh-Otwani**
Working towards closing #93 I've added tests for intersections and unions for sets.

[Remark: We should probably add tests for `empty` and maybe I should just do that in the PR but it looks obviously correct to me. (Saying this, I feel like I'm going to end up having to add it.)]

Issue #93 explains that the mutable data test suite is meant to provide properties which serve as axioms that define the behavior of how to interact with the data structure. That is, if all the properties hold, the behavior of the data structure is mathematically specified. For each accessor function `f` and each modifier or constructor `g` we have one axiom to specify the behavior of `f (g (...)) = ...`. See for instance [this](https://softwarefoundations.cis.upenn.edu/vfa-current/ADT.html) link. 

In this PR, since we added two modifiers (intersection and union) and have two basic accessors (member and size), we need 2 X 2 = 4 axioms, which are precisely the 4 tests I've added.

### CODE REVIEW

**Comment by Divesh-Otwani** [comment]
Also, no idea why the CI is failing since `nix-shell --pure --run 'cabal update; cabal test'` worked on my machine.

**Comment by facundominguez** [comment]
Retrying the CI job succeeds. Maybe there is a full disk in the machine that ran the first attempt.

File: test/Test/Data/Mutable/Set.hs [hunk 1]
```diff
@@ -129,9 +141,39 @@ memberDelete2Test val1 val2 set = fromRead (Set.member val2 set)
...
+  v <- forAll value
+  let valInBoth = elem v l && elem v l'
+  test $ unur Linear.$
+    Set.fromList l
+      (\s -> Set.fromList l'
+        (\s' -> memberIntersectTest v valInBoth (Set.intersection s s')))
+
+memberIntersectTest :: Int -> Bool -> Set.Set Int %1-> Ur (TestT IO ())
+memberIntersectTest val result inter = Set.member val inter Linear.&
+  \(Ur b, s') -> lseq s' (Ur (b === result))
```
Comment (aspiwack) [inline test/Test/Data/Mutable/Set.hs:172]: This test is correct. But I contend that it doesn't test much, as I assume that, almost always, `v` will not be a member of either `l` or `l'` (let alone both). Therefore, this test will almost always be testing that `False === False`.

Instead, I'd suggest writing a `toList` function of sort, and testing that the output list is the expected one. Individual tests will be a bit more expensive, but much more informative.

File: test/Test/Data/Mutable/Set.hs [hunk 2]
```diff
@@ -192,3 +234,31 @@ sizeDelete2Test val set = fromRead (Set.size set)
...
+
+sizeIntersect :: Property
+sizeIntersect = property $ do
+  l <- forAll list
+  l' <- forAll list
+  let isectListLen = List.length $ List.nub $ List.intersect l l'
+  let f :: Set.Set Int %1-> Ur Int
+      f = \s -> Set.fromList l' (\s' -> dropSnd (Set.size (Set.intersection s s')))
+  let isectInSetsLen = unur Linear.$ Set.fromList l f
+  test (isectInSetsLen === isectListLen)
```
Comment (aspiwack) [inline test/Test/Data/Mutable/Set.hs:256]: Same comment.

File: test/Test/Data/Mutable/Set.hs (hunk 1 above)
Comment (Divesh-Otwani) [reply test/Test/Data/Mutable/Set.hs]: Ah, so pseudo-homomorphism tests. It's not trivial to see, but these are sufficient to replace defining axioms. Done below.

File: test/Test/Data/Mutable/Set.hs [hunk 3]
```diff
@@ -192,3 +221,38 @@ sizeDelete2Test val set = fromRead (Set.size set)
...
+toListFromList = property $ do
+  l <- forAll list
+  let outsideSet = List.nub . List.sort $ l
+  List.sort (unur (Set.fromList l Set.toList)) === outsideSet
+
+unionHomomorphism :: Property
+unionHomomorphism = property $ do
+  l <- forAll list
+  l' <- forAll list
+  let listUnion = List.nub $ List.sort $ List.union l l'
```
Comment (utdemir) [inline test/Test/Data/Mutable/Set.hs:235]: `nub` is pretty inefficient (`O(n^2)`), but we should be able to replace it with `Data.Containers.ListUtils.nubOrd` which runs in `O(nlogn)`. It doesn't change much, but will probably make the tests a bit faster.

Comment (Divesh-Otwani) [reply test/Test/Data/Mutable/Set.hs:235]: Good idea :+1:

**Comment by Divesh-Otwani** [comment]
I cleaned up those tests and convinced myself of a way to still test the axioms if we are using "homomorphism" tests against reference list implementations.

File: test/Test/Data/Mutable/Set.hs [hunk 4]
```diff
@@ -2,18 +2,56 @@
...
+-- If however, this is cumbersome, we can often just test against an existing
+-- implementation. This is like a homomorphism test (but it's not quite a
+-- homomorphism). For unions, this is like saying if we have two lists A and B,
+-- and we take their union as lists, then sort and nub (remove duplicates)
+-- this is the same as using @Set.fromList@ to make them into sets, taking
+-- the union as sets, converting back with @Set.toList@ and then sorting
+-- the output list.
+--
+-- To have such a test replace an axiom, however, we need to have "homomorphisms"
+-- for each accessor. E.g., for all x and l, @member x (fromList l) = elem x l@.
```
Comment (aspiwack) [inline test/Test/Data/Mutable/Set.hs:24]: Like in #265: use `toList`, rather than `fromList`.

And probably phrase this in terms of semantics, rather than homomorphism.

You can spare yourself the cumbersomeness of using lists in your semantics if you use `Set` instead.

Comment (Divesh-Otwani) [reply test/Test/Data/Mutable/Set.hs:24]: I like homomorphisms better though :cry: It's like saying the semantics of operations on sets can be expressed in terms of the semantics of operations on lists because they can carry through the `fromList`. I'll add a comment stating this though.

Comment (Divesh-Otwani) [reply test/Test/Data/Mutable/Set.hs]: Also, I want to use lists in my semantics because we don't have a `fromSet` constructor and it won't correspond as well. (Though we easily could have such a constructor.)

**Review by aspiwack — APPROVED** [review]
I still think this homorphism thing is a bit misleading.

The semantics of a `Set` is, well, a (finite) set. And the semantic of `union` on `Set` is the union on the semantics. Then we get to claim that the semantics is an homomorphism for the `union` structure, which, technically, is true. But we are not testing the semantics, we are testing the union.

That being said, there is a way in which we can legitimately speak of homomorphism. Let's imagine that we have an abstract data type. It has a semantics and all. But, in this view, the semantics is not executable. However we use `[]` (or `Data.Set`) as a _reference implementation_. Then, we establish correctness of our implementation of interest by exhibiting a (semantic-preserving) homomorphism from our structure to the reference implementation.
